ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/PullPhysicalPlan.java

```diff
@@ -123,26 +123,16 @@ public Materialization getMaterialization() {
   }
 
   public List<KsqlKey> getKeys() {
-    if (requiresRequestsToAllPartitions()) {
-      return Collections.emptyList();
+    final List<KsqlKey> list = new ArrayList<>();
+    for (LookupConstraint c : lookupConstraints) {
+      if (c instanceof KeyConstraint) {
+        final KeyConstraint kc = (KeyConstraint) c;
+        list.add(kc.getKsqlKey());
+      } else {
+        return Collections.emptyList();
+      }
     }
-    return lookupConstraints.stream()
-        .filter(lookupConstraint -> lookupConstraint instanceof KeyConstraint)
-        .map(KeyConstraint.class::cast)
-        .filter(keyConstraint -> keyConstraint.getConstraintOperator() == ConstraintOperator.EQUAL)
-        .map(KeyConstraint::getKsqlKey)
-        .collect(ImmutableList.toImmutableList());
-  }
-
-  private boolean requiresRequestsToAllPartitions() {
-    return lookupConstraints.stream()
-        .anyMatch(lookupConstraint -> {
-          if (lookupConstraint instanceof KeyConstraint) {
-            final KeyConstraint keyConstraint = (KeyConstraint) lookupConstraint;
-            return keyConstraint.getConstraintOperator() != ConstraintOperator.EQUAL;
-          }
-          return true;
-        });
+    return ImmutableList.copyOf(list);
```

Comment:
It's possible if you have a query like `(KEY < 'abc' AND KEY > 'a') OR (KEY > 'def' AND KEY < 'e') OR (KEY <= `z` AND KEY >= 'y'`), you'll send up doing three scans which might be more expensive than doing a single table scans.  This is where a query optimizer could eventually be smart.  I think it's fine to do this at the moment, but it might make sense to try a benchmark at some point for some representative queries to see what's faster.

Also, at the moment, we do some deduping for the request for queries like `(KEY = 'a' OR KEY = 'a')` so that it will return a single row and won't issue the same query twice (even though that would be a nonsensical query).  It's possible to issue a query like `KEY < 'abc' OR KEY < 'a'`, which could also result in duplicates, though I don't think we'd dedupe the requests since they're a bit different.  The options are to:
- Recognize ranges and optimize the above query to be just `KEY < 'a'` -- this is something that would be nice to do, though it probably shouldn't be part of this PR.
- Dedupe results, not just queries.  This is a bit hard because we stream back results at the moment and are somewhat stateless.  This would require keeping a set of keys we've returned -- not a bad idea, though it means we could run out of memory or use a lot of it with a sufficiently large table size.  Probably a no-go for our use cases.
- Allow duplicates. I think the SQL purists would probably be angry here since this would break the model.
- Do a single table scan if you see more than one range query.  This works without issue, but might not be as performant as a few small double-bounded range queries.

I'm for the last solution since it's the simplest.  Curious what you think.

Reply:
Yea, the worst would be to have multiple range scans for such multi-range queries, the best would be to optimize those queries and compact the expressions. As for now the plan is to defer the optimization/compaction to a separate PR and fallback to table scan for multi-range queries so that at least we are not any worse than without this PR.

Reply:
Sounds good to me.

I think the code at the moment doesn't yet do this, if I'm not mistaken.  We'd probably want to do a check at the logical layer in QueryFilterNode.  I made a suggestion about that below.

Reply:
I think the current code handles `KEY < 'a' AND KEY > 'b'`, allowing this to be a table scan, but currently, `KEY < 'a' OR KEY > 'b'` would result is two separate range queries and some of the issues I was referring to requiring deduping.

Reply:
Queries with multiple disjuncts will either be key lookups if all disjuncts are Equality, or table scans. Based on how we determine the plan in PullPhysicalPlanBuilder. Validation tests should cover this.